Log Fragmentation

martinsumner opened the pull request:
With the leveled backend, memory fragmentation is an ongoing concern.  So, here the regular compaction callback is used to log information about carrier sizes and carrier block sizes for key allocators.

martinsumner committed:
Log Fragmentation

With the leveled backend, memory fragmentation is an ongoing concern.  So, here the regular compaction callback is used to log information about carrier sizes and carrier block sizes for key allocators.

ThomasArts commented on src/riak_kv_leveled_backend.erl:
```diff
@@ -635,6 +637,31 @@ callback(Ref, UnexpectedCallback, State) ->
+    FragStats =
+        lists:filter(
+            fun(MemStuff) ->
+                element(1, element(1, MemStuff)) == Allocator
```
This is a bit risky.. assume data type changes in the future, then this may actually crash as a side effect of logging.
Rather make 3 match cases:
lists:filter(fun({{A, _}, _Stats}) -> A == Allocator;
                       (_) -> false
                       end,

ThomasArts replied:
alternatively, build the new list straigth away:
FragStats = [ Stats || {{A, _}, Stats} <- econ_alloc:fragmentation(current), A == Allocator ]

ThomasArts commented on src/riak_kv_leveled_backend.erl:
```diff
@@ -635,6 +637,31 @@ callback(Ref, UnexpectedCallback, State) ->
+            recon_alloc:fragmentation(current)),
+    {MB_BS, MB_CS, SB_BS, SB_CS} =
+        lists:foldl(
+            fun({{_A, _I}, Stats}, {MBAcc, MCAcc, SBAcc, SCAcc}) ->
```
or here 
```suggestion
            fun({{A, _I}, Stats}, {MBAcc, MCAcc, SBAcc, SCAcc}) when A == Allocator->
```
and a separte clause for when it's not.

martinsumner committed:
Correct log format

martinsumner committed:
Update riak_kv_leveled_backend.erl

martinsumner committed:
Update to account for review comments

martinsumner committed:
Spawn function to log

There may be a very large number of allocators, so the cost of calling recon_alloc:fragmentation/1 and processing the outputs could be high - so don't lock the vnode in this case.

martinsumner committed:
Don't make log function depend on format of recon_alloc:fragmentation/1

martinsumner committed:
Update rebar.config
